ann_hw/src/neuralnet.py

```diff
@@ -0,0 +1,119 @@
+            test = strat_sets.pop(f)
+            tests[f] = test
+            for fold, fold_set in strat_sets.iteritems():
+                for instance, row in fold_set.iterrows():
+                ####################
+                ### feed forward ###
+                ####################
+                    # input layer output: o_i
+                    y_d, o_i = get_nn_input(row, instance, classes)
+                    #net_ij = np.dot(np.transpose(o_i), w_ij).tolist()
+                    net_ij = np.dot(o_i, w_ij).tolist()
+                    # hidden layer output: o_j
+                    o_j = get_nn_hidden(sigmoid(net_ij), instance)
+                    net_jk = np.dot(o_j, w_jk)
+                    # output node: o_k
+                    o_k = sigmoid(net_jk)
+                #####################
+                ### backpropogate ###
+                #####################
+                    # update hidden weights
+                    d_k = (y_d - o_k) * o_k * (1 - o_k)
+                    delta_jk = learning_rate * d_k * np.array([o_j])
+                    w_jk = np.transpose(delta_jk) + w_jk
+                    # update input weights
+                    d_j = d_k * np.sum(w_jk, 1) * (1 - np.array([o_j])) * np.array([o_j])
+                    delta_ij = np.dot(np.transpose([o_i]), learning_rate * d_j)
+                    delta_ij = np.delete(delta_ij, -1, 1) # hidden bias doesn't connect to input layer
+                    w_ij = delta_ij + w_ij
+    return w_ij, w_jk, tests, classes
+
+
+def predict(w_ij, w_jk, tests, classes):
+    """
+    Taking the model from the training phase (i.e. the weights for each 
+    node in each layer), this uses the test set to calculate a prediction
+    and the "confidence" (sigmoid value) of that prediction.
+    """
+    to_print = []
+    for fold, test_set in tests.iteritems():
+        for instance, row in test_set.iterrows():
+            y_d, o_i = get_nn_input(row, instance, classes)
+            net_ij = np.dot(o_i, w_ij).tolist()
+            o_j = get_nn_hidden(sigmoid(net_ij), instance)
+            net_jk = np.dot(o_j, w_jk)
+            o_k = sigmoid(net_jk)[0]
+            if o_k >= 0.5:
+                o_d = 1
+            else:
+                o_d = 0
+            line = (instance+1, fold, o_d, y_d, o_k)
+            to_print.append(line)
+    return to_print
+
+def main():
+    """
+    This neural net algorithm builds an ANN for training data that is 
+    real-valued for every attribute. It is written for a single hidden layer of 
+    equal size to the input layer, a bias node on each, and assumes binary 
+    classification.
+    """
+    # get input: training data
+    train = arff.load(open(sys.argv[1], 'rb')) 
+    num_folds = int(sys.argv[2])
+    learning_rate = float(sys.argv[3])
+    num_epochs = int(sys.argv[4])
+    
+    # data and attributes
+    attr = pd.DataFrame(train['attributes'], columns=['attr', 'vars'])
```

Comment:
I struggled with pandas manipulation so there's "to-list's" and whatnot hanging around.